python/plugins/processing/algs/qgis/CheckValidity.py

```diff
@@ -0,0 +1,182 @@
+
+"""
+***************************************************************************
+    CheckValidity.py
+    ---------------------
+    Date                 : May 2015
+    Copyright            : (C) 2015 by Arnaud Morvan
+    Email                : arnaud dot morvan at camptocamp dot com
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Arnaud Morvan'
+__date__ = 'May 2015'
+__copyright__ = '(C) 2015, Arnaud Morvan'
+
+# This will get replaced with a git SHA1 when you do a git archive323
+
+__revision__ = '$Format:%H$'
+
+from PyQt4 import QtCore
+from qgis.core import QGis, QgsGeometry, QgsFeature, QgsField
+from processing.core.GeoAlgorithm import GeoAlgorithm
+from processing.core.parameters import ParameterVector
+from processing.core.parameters import ParameterSelection
+from processing.core.outputs import OutputVector
+from processing.tools import dataobjects, vector
+
+settings_method_key = "/qgis/digitizing/validate_geometries"
+
+
+class CheckValidity(GeoAlgorithm):
+
+    INPUT_LAYER = 'INPUT_LAYER'
+    METHOD = 'METHOD'
+    VALID_OUTPUT = 'VALID_OUTPUT'
+    INVALID_OUTPUT = 'INVALID_OUTPUT'
+    ERROR_OUTPUT = 'ERROR_OUTPUT'
+
+    METHODS = ['The one selected in digitizing settings',
+               'QGIS',
+               'GEOS']
+
+    def defineCharacteristics(self):
+        self.name = 'Check validity'
+        self.group = 'Vector geometry tools'
+
+        self.addParameter(ParameterVector(
+            self.INPUT_LAYER,
+            self.tr('Input layer'),
+            [ParameterVector.VECTOR_TYPE_ANY]))
+
+        self.addParameter(ParameterSelection(
+            self.METHOD,
+            self.tr('Method'),
+            self.METHODS))
+
+        self.addOutput(OutputVector(
+            self.VALID_OUTPUT,
+            self.tr('Valid output')))
+
+        self.addOutput(OutputVector(
+            self.INVALID_OUTPUT,
+            self.tr('Invalid output')))
+
+        self.addOutput(OutputVector(
+            self.ERROR_OUTPUT,
+            self.tr('Error output')))
+
+    def processAlgorithm(self, progress):
+        settings = QtCore.QSettings()
+        initial_method_setting = settings.value(settings_method_key, 1)
+
+        method = self.getParameterValue(self.METHOD)
+        if method != 0:
+            settings.setValue(settings_method_key, method)
+        try:
+            self.doCheck(progress)
+        except:
+            raise
```

Comment:
Not sure about that, a GeoAlgorithmExecutionException could be raised for example to signal incompatible input values, something specific to the algorithm.
Here the purpose is to resend the original exception that was raised (maybe a GeoAlgorithmExecutionException but maybe not) and to finally be sure to restore the setting value in each case.
Raising an GeoAlgorithmExecutionException here would mislead the caller on the original exception type.
If the caller want to catch all exception types and display them to the user, he can do so.